bpf/lib/common.h

```diff
@@ -264,6 +264,7 @@ struct edt_info {
 struct remote_endpoint_info {
 	__u32		sec_label;
 	__u32		tunnel_endpoint;
+	mac_t		vtep_mac;
```

Comment:
I find another way to address this issue, but need input. when this feature is enabled,  we could create macro in `node_config.h`, and at run time, the BPF code could take the `vtep_mac` from macro in `node_config.h`, just like the `WORLD_ID` that is defined in `node_config.h`, this way, we don't  need to add `vtep_mac` field in `remote_endpoint_info`, a working diff like below:

```
diff --git a/bpf/bpf_lxc.c b/bpf/bpf_lxc.c
index b5e852e20..41cf68199 100644
--- a/bpf/bpf_lxc.c
+++ b/bpf/bpf_lxc.c
@@ -624,7 +624,7 @@ skip_service_lookup:
                        tunnel_endpoint = info->tunnel_endpoint;
                        encrypt_key = get_min_encrypt_key(info->key);
 #ifdef ENABLE_VTEP
-                       vtep_mac = info->vtep_mac;
+                       vtep_mac = VTEP_MAC;
 #endif
 #ifdef ENABLE_WIREGUARD
                        /* If we detect that the dst is a remote endpoint, we
@@ -822,7 +822,7 @@ skip_egress_gateway:
 
 #ifdef ENABLE_VTEP
        {
-               if (vtep_mac && tunnel_endpoint) {
+               if (vtep_mac && (tunnel_endpoint == VTEP_ENDPOINT)) {
                        if (eth_store_daddr(ctx, (__u8 *)&vtep_mac, 0) < 0)
                                return DROP_WRITE_ERROR;
                        return __encap_and_redirect_with_nodeid(ctx, tunnel_endpoint,
diff --git a/pkg/datapath/linux/config/config.go b/pkg/datapath/linux/config/config.go
index 679eb6231..2a2d2cd06 100644
--- a/pkg/datapath/linux/config/config.go
+++ b/pkg/datapath/linux/config/config.go
@@ -7,6 +7,7 @@ import (
        "bufio"
        "bytes"
        "encoding/base64"
+       "encoding/binary"
        "encoding/json"
        "fmt"
        "io"
@@ -537,6 +538,16 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 
        if option.Config.EnableVTEP {
                cDefinesMap["ENABLE_VTEP"] = "1"
+
+               ip := option.Config.VtepEndpoints[0]
+               cDefinesMap["VTEP_ENDPOINT"] = fmt.Sprintf("%#x", byteorder.NetIPv4ToHost32(ip))
+
+               mac := option.Config.VtepMACs[0]
+               //append two bytes to avoid binary LittleEndian error complain
+               zero := []byte{0x00, 0x00}
+               vtep_mac := append(mac, zero...)
+               num := binary.LittleEndian.Uint64(vtep_mac)
+               cDefinesMap["VTEP_MAC"] = fmt.Sprintf("%#x", num)
        }

``` 
the vtep config:

```
  enable-vtep:   "true"
  vtep-endpoint: "10.169.72.236"
  vtep-cidr:     "10.1.5.0/24"
  vtep-mac:      "82:36:4c:98:2e:56"

```

the macro result in `node_config.h` at run time.

```
/var/run/cilium/state/globals/node_config.h:#define ENABLE_VTEP 1
/var/run/cilium/state/globals/node_config.h:#define VTEP_ENDPOINT 0xec48a90a
/var/run/cilium/state/globals/node_config.h:#define VTEP_MAC 0x562e984c3682
```

now the remaining issue is what if I have multiple VTEPs needs to be defined in `node_config.h`, sorry send out this message at weekend, want to write down my thought @joestringer @pchaigno any suggestions would be appreciated when you have time :)

Reply:
Indeed, this can work for one VTEP but it's less clear how you expand this approach to support multiple VTEPs. Ultimately I don't think we want to introduce linear iteration in a path like this to handle VTEPs, both because this increases the latency for handling packets and also the total instruction complexity count which is a limited resource. The obvious next step after that is to abstract the lookup into a map... which is kinda what the previous tunnel_map / ipcache extension approaches were trying to do.

Maybe there's another option there where instead of extending one of the existing maps, there's a brand new map particularly for VTEP support. Then the memory increase can be bounded quite nicely to the expected number of VTEPs and that could be configured up front. I haven't thought too much about this yet though.

Reply:
the diff below is what I come up to  resolve multiple VTEP issues, basically to add array macro for VTEP devices in `node_config.h`, it seems a bit of hack.  there is bounded for loop, I can check how much extra instructions the loop added, with this approach,   when this feature is disabled by default, it does not add extra instructions complexity or memory usage, when originally started this feature,  it is aimed for BIG-IP which usually only require two VTEPs for high availability.  I think it is good enough for people to try :), we can improve later if this feature really brings lot of interests

```
diff --git a/bpf/bpf_lxc.c b/bpf/bpf_lxc.c
index b5e852e20..5b0a423c4 100644
--- a/bpf/bpf_lxc.c
+++ b/bpf/bpf_lxc.c
@@ -623,9 +623,6 @@ skip_service_lookup:
                        *dst_id = info->sec_label;
                        tunnel_endpoint = info->tunnel_endpoint;
                        encrypt_key = get_min_encrypt_key(info->key);
-#ifdef ENABLE_VTEP
-                       vtep_mac = info->vtep_mac;
-#endif
 #ifdef ENABLE_WIREGUARD
                        /* If we detect that the dst is a remote endpoint, we
                         * need to mark the packet. The ip rule which matches
@@ -822,6 +819,14 @@ skip_egress_gateway:
 
 #ifdef ENABLE_VTEP
        {
+               int i;
+               for (i = 0; i < VTEP_NUMS; i++) {
+                       if (tunnel_endpoint == VTEP_ENDPOINT[i]) {
+                               vtep_mac = VTEP_MAC[i];
+                               break;
+                       }
+               }
+
                if (vtep_mac && tunnel_endpoint) {
                        if (eth_store_daddr(ctx, (__u8 *)&vtep_mac, 0) < 0)
                                return DROP_WRITE_ERROR;

diff --git a/bpf/lib/common.h b/bpf/lib/common.h
index a82ededa1..ee12e37c7 100644
--- a/bpf/lib/common.h
+++ b/bpf/lib/common.h
@@ -264,7 +264,6 @@ struct edt_info {
 struct remote_endpoint_info {
        __u32           sec_label;
        __u32           tunnel_endpoint;
-       mac_t           vtep_mac;
        __u8            key;
 };

diff --git a/daemon/cmd/datapath.go b/daemon/cmd/datapath.go
index 0d4e907cf..4cde300ff 100644
--- a/daemon/cmd/datapath.go
+++ b/daemon/cmd/datapath.go
@@ -474,7 +474,7 @@ func setupIPCacheVTEPMapping() error {
                        logfields.IPAddr: ep,
                }).Debug("Updating ipcache map entry for VTEP")
 
-               err := ipcachemap.UpdateIPCacheMapping(option.Config.VtepCIDRs[i], ep, vtepID, option.Config.VtepMACs[i], encryptKey)
+               err := ipcachemap.UpdateIPCacheMapping(option.Config.VtepCIDRs[i], ep, vtepID, encryptKey)
                if err != nil {
                        return fmt.Errorf("Unable to set up VTEP ipcache mappings: %w", err)
                }

diff --git a/pkg/datapath/linux/config/config.go b/pkg/datapath/linux/config/config.go
index 679eb6231..8514ad8f6 100644
--- a/pkg/datapath/linux/config/config.go
+++ b/pkg/datapath/linux/config/config.go
@@ -7,11 +7,13 @@ import (
        "bufio"
        "bytes"
        "encoding/base64"
+       "encoding/binary"
        "encoding/json"
        "fmt"
        "io"
        "net"
        "sort"
+       "strconv"
        "text/template"
 
        "github.com/cilium/cilium/pkg/bpf"
@@ -537,6 +539,28 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 
        if option.Config.EnableVTEP {
                cDefinesMap["ENABLE_VTEP"] = "1"
+
+               l := len(option.Config.VtepEndpoints)
+               cDefinesMap["VTEP_NUMS"] = strconv.Itoa(l)
+
+               ip_str := "(__u32[]){"
+               mac_str := "(__u64[]){"
+
+               for i, ep := range option.Config.VtepEndpoints {
+
+                       ip_str = ip_str + fmt.Sprintf("%#x", byteorder.NetIPv4ToHost32(ep)) + ", "
+
+                       mac := option.Config.VtepMACs[i]
+                       //append two bytes to avoid binary LittleEndian error complain
+                       zero := []byte{0x00, 0x00}
+                       vtep_mac := append(mac, zero...)
+                       num := binary.LittleEndian.Uint64(vtep_mac)
+                       mac_str = mac_str + fmt.Sprintf("%#x", num) + ", "
+
+               }
+               cDefinesMap["VTEP_ENDPOINT"] = ip_str + "}"
+               cDefinesMap["VTEP_MAC"] = mac_str + "}"
+
        }
 
        vlanFilter, err := vlanFilterMacros()

diff --git a/pkg/maps/ipcache/ipcache.go b/pkg/maps/ipcache/ipcache.go
index 48f2cd917..a8a1a025b 100644
--- a/pkg/maps/ipcache/ipcache.go
+++ b/pkg/maps/ipcache/ipcache.go
@@ -14,7 +14,6 @@ import (
        "github.com/cilium/cilium/pkg/cidr"
        "github.com/cilium/cilium/pkg/logging"
        "github.com/cilium/cilium/pkg/logging/logfields"
-       "github.com/cilium/cilium/pkg/mac"
        "github.com/cilium/cilium/pkg/types"
 
        "golang.org/x/sys/unix"
@@ -122,15 +121,14 @@ func NewKey(ip net.IP, mask net.IPMask) Key {
 // +k8s:deepcopy-gen=true
 // +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapValue
 type RemoteEndpointInfo struct {
-       SecurityIdentity uint32        `align:"sec_label"`
-       TunnelEndpoint   types.IPv4    `align:"tunnel_endpoint"`
-       VtepMAC          mac.Uint64MAC `align:"vtep_mac"`
-       Key              uint8         `align:"key"`
+       SecurityIdentity uint32     `align:"sec_label"`
+       TunnelEndpoint   types.IPv4 `align:"tunnel_endpoint"`
+       Key              uint8      `align:"key"`
 }
 
 func (v *RemoteEndpointInfo) String() string {
-       return fmt.Sprintf("identity=%d encryptkey=%d tunnelendpoint=%s vtepmac=%s",
-               v.SecurityIdentity, v.Key, v.TunnelEndpoint, v.VtepMAC)
+       return fmt.Sprintf("identity=%d encryptkey=%d tunnelendpoint=%s",
+               v.SecurityIdentity, v.Key, v.TunnelEndpoint)
 }
 
 // GetValuePtr returns the unsafe pointer to the BPF value.
@@ -255,18 +253,12 @@ func Reopen() error {
 
 // Function to update IPCache map with node PodCIDR, VTEP CIDR
 func UpdateIPCacheMapping(newCIDR *cidr.CIDR, newTunnelEndpoint net.IP,
-       securityIdentity uint32, vtepMAC mac.MAC, encryptKey uint8) error {
+       securityIdentity uint32, encryptKey uint8) error {
 
        key := NewKey(newCIDR.IP, newCIDR.Mask)
 
-       mac, err := vtepMAC.Uint64()
-       if err != nil {
-               return err
-       }
-
        value := RemoteEndpointInfo{
                SecurityIdentity: securityIdentity,
-               VtepMAC:          mac,
                Key:              encryptKey,
        }
        if ip4 := newTunnelEndpoint.To4(); ip4 != nil {
```

Reply:
> the diff below is what I come up to resolve multiple VTEP issues, basically to add array macro for VTEP devices in `node_config.h`, it seems a bit of hack. there is bounded for loop, I can check how much extra instructions the loop added, with this approach, when this feature is disabled by default, it does not add extra instructions complexity or memory usage, when originally started this feature, it is aimed for BIG-IP which usually only require two VTEPs for high availability. I think it is good enough for people to try :), we can improve later if this feature really brings lot of interests

when this feature is enabled, with the bounded for loop, it added extra `22` instructions

```
[root@centos-dev home]# llvm-objdump -S /home/noloop.o --section="2/7"  > noloop.txt

[root@centos-dev home]# llvm-objdump -S /home/loop.o --section="2/7"  > loop.txt

[root@centos-dev home]# tail -1 loop.txt 
    4201:	05 00 9c f4 00 00 00 00	goto -2916 <LBB1_163>
[root@centos-dev home]# tail -1 noloop.txt 
    4179:	05 00 b2 f4 00 00 00 00	goto -2894 <LBB1_163>

```

Reply:
interesting yeah. If the limit really is 2 and we can test with datapath configurations enabling this functionality, then I think I'm OK with this solution. The concern I would have here is not so much the absolute number of 22 additional instructions, but the way that these 22 additional instructions may influence the Linux verifier's assessment of complexity. If Cilium is generating programs that are at the limits of what the Linux verifier will accept and we add 22 additional instructions, then that can tip the whole program over the limit and cause the entire datapath to fail to load. I do not want users to be the first ones to find this out when they configure some specific configuration that happens to increase the overall complexity; I'd like to see us try to find those limits in CI and put guardrails in place to avoid forcing users to find this out the hard way.

Reply:
@joestringer I rebased on the master and made the new changes to add array macro for VTEP devices in `node_config.h`, please review when you have time.